File: src/openrc-firstboot.py
#!/usr/bin/env python3
"""
SPDX-License-Identifier: MIT

openrc-firstboot: A very, very, *very* simple firstboot in Python.
                  Asks the user to set a hostname and root password, then lets
                  them create a local user. Removes itself from the 'default'
                  runlevel, but does not destroy the init script.

(C) 2021 James Calligeros <jcalligeros99@gmail.com>
"""

from crypt import crypt
from getpass import getpass
from random import choice
from subprocess import call
from time import sleep


def set_password(user, newpasswd):
    """
    Change password for 'user' to 'newpasswd'
    """
    # Set some characters to use as salt for encrypting the password
    SALT_DICT = ("0123456789",
                 "ABCDEFGHIJKLMNOPQRSTUVWXYZ",
                 "abcdefghijklmnopqrstuvwxyz")

    salt = "".join(choice(ALPHABET) for i in range(8))
    mask = crypt(newpasswd, "$1$"+salt+"$")

    ret = call(("usermod", "-p", mask, user))
    if ret != 0:
        print(f"Error setting password for {user}.")
    else:
        print(f"Password for {user} changed successfully.")


def set_hostname(newname):
    """
    Opens/creates /etc/hostname and dumps the chosen hostname in.
    """

    try:
        with open("/etc/hostname", mode="w") as hn:
            hn.write(newname)
    except:
        print("Unable to set the hostname for the machine.")
    else:
        print(f"This machine will now be known as {newname} on the network.")


def create_local_user(username):
    """
    Creates a local user and adds them to the audio and video groups.
    """
    ret = call(("usermod",
                "-a -G audio,video,wheel",
                username,
                "-s /bin/bash"))
    if ret != 0:
        print(f"Unable to create user {username}.")
    else:
        print((f"New user {username} created successfully and ",
               "added to the 'audio', 'video' and 'wheel' groups. ",
               "Your default shell is bash."))

def cleanup():
    """
    Removes the init script from the default runlevel and reboots the machine.
    """
    call("rc-update delete openrc-firstboot default")
    call("reboot")


def main():
    print("#######################")
    print("FIRST BOOT SYSTEM SETUP")
    print("#######################")

    print("This appears to be your first time booting this system. This ",
          "script will guide you through: ")
    print("1. Changing the default root password.")
    print("2. Setting a hostname for the machine.")
    print("3. Creating a local user.\n\n\n\n")
    sleep(3)

    print("##################################")
    print("Changing the default root password")
    print("##################################\n\n")
    while True:
        newroot = getpass("Please input a new root password.")
        conf    = getpass("Confirm root password: ")
        if newroot != conf:
            print("Error: passwords do not match. Try again.")
            sleep(2)
            continue
        else:
            set_password("root", newroot)
            break

    print("##################################")
    print("Setting a hostname for the machine")
    print("##################################\n\n")
    print("The hostname identifies your machine on the network. Must contain ",
          "alphanumeric characters only. Do not append '.local'.\n")
    name = input("Please specify a hostname: ")
    set_hostname(name)

    print("#####################")
    print("Creating a local user")
    print("#####################\n\n")
    print("You will now create a user account for yourself.\n")
    uname = input("Please enter a username: ")
        while True:
        upass   = getpass(f"Please input a password for {name}.")
        conf    = getpass("Confirm password: ")
        if upass != conf:
            print("Error: passwords do not match. Try again.")
            sleep(2)
            continue
        else:
            set_password(name, upass)
            break

    print("Your machine will now reboot, probably to a login shell. Consult ",
          "your distro's documentation for further configuration and ",
          "customisation options. This script will not run again by default, ",
          "but can be made to by running\n")
    print("rc-update add openrc-firstboot default\n")
    print("as root and then rebooting your machine.")

    input("Press Enter to reboot... ")

    cleanup()


if __name__ == "__main__":
    main()

File: src/openrc-firstboot.c
/* SPDX-License-Identifier: MIT
 *
 * openrc-firstboot
 *
 * An openrc-friendly implementation of functionality similar to
 * systemd-firstboot.
 *
 * this is all awful btw and needs a LOT of reworking
 *
 * (C) 2021 James Calligeros <jcalligeros99@gmail.com>
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>


void set_root_password()
{
	int status;


	status = system("passwd root");
	if (status != 0)
	{
		printf("Unable to set password for root.\n");
		exit(EXIT_FAILURE);
	} else {
		printf("Password for root set successfully.\n");
		sleep(2);
	}
}

void set_hostname()
{
	char hostname;
	FILE *etc_hostname;

	printf("Please enter a hostname: ");
	scanf("%s", &hostname);

	etc_hostname = fopen("/etc/hostname", "w");
	fprintf(etc_hostname, "%s\n", &hostname);
	fclose(etc_hostname);
}


void create_user()
/* this is psychotic C butchery. pls fix :( */
{
	int status;
    const char usrcmd;
    const char pwdcmd;
	usrcmd = "useradd -m -G audio,video,wheel -s /bin/bash ";
	pwdcmd = "passwd ";
    char username;
    printf("Please specify a username: ");
    scanf("%s", username);

    char* mkuser = (char *)malloc(strlen(usrcmd) + strlen(username));
    strcpy(mkuser, usrcmd);
    strcat(mkuser, username);

    char* mkpswd = (char *)malloc(strlen(pwdcmd) + strlen(username));
    strcpy(mkpswd, pwdcmd);
    strcat(mkpswd, username);


    sleep(0.2);
	printf("%s\n", mkuser); /* debug - replace with a system() call */
	if (status != 0)
	{
		printf("Unable to create user %s.\n", username);
		exit(EXIT_FAILURE);
	} else {
		printf("Now, specify a password for %s\n", username);
		printf("%s\n", mkpswd); /* debug - replace with a system() call */
	}

	return 0;
}



void cleanup()
{
	int status;
	status = system("rc-update delete firstboot default");
	if (status != 0)
	{
		printf("Unable to remove firstboot from the runlevel.");
		exit(EXIT_FAILURE);
	} else {
		system("reboot");
	}
}


int main()
{
	printf("################\n");
	printf("openrc-firstboot\n");
	printf("################\n\n");

	printf("This appears to be your first time booting this system. ");
	printf("You will now be guided through:\n");
	printf("1. Changing the root password\n");
	printf("2. Setting a hostname for the machine\n");
	printf("3. Creating a local user\n\n");
	sleep(8);

	printf("##########################\n");
	printf("Changing the root password\n");
	printf("##########################\n\n");
	printf("You will now change the root password. ARM distro images ");
	printf("often set insecure default root passwords. It is therefore\n");
	printf("in your best interests to change this.\n\n");

	set_root_password();



	printf("####################\n");
	printf("Setting the hostname\n");
	printf("####################\n\n");

	printf("Please specify an alphanumeric hostname for the machine.\n");
	printf("Do not append '.local' or any special characters.\n\n");

	set_hostname();



	printf("#####################\n");
	printf("Creating a local user\n");
	printf("#####################\n\n");

	printf("You will now create a local user account.\n");

	create_user();

	printf("Your machine will now reboot, probably to a login shell. ");
	printf("Please consult your\n distro's documentation for futher ");
	printf("configuration and customisation options.\nThis script will ");
	printf("not run again by default, but can be made to do so by ");
	printf("running\n\nrc-update add firstboot\n\nand rebooting.\n\n");

	printf("Rebooting in 10 seconds...\n");
	sleep(10);

	cleanup();

	return 0;
}
